[lit] Echo full RUN lines in case of external shells

jdenny-ornl opened the pull request:
**NOTE: The first two commits in this PR (a398f3e1dfbfc83489e4d85fd74e8b3f12d469e9 and e4c20b93fa8d49ac22f4493358a8e3c8102eaf51) should not be reviewed here.  They contain [D154984](https://reviews.llvm.org/D154984) and [D156954](https://reviews.llvm.org/D156954), which will be re-landed at the same time as this PR.  Please review the third commit (d71c72c9f20ed34dc2fe2d0ec849a6e44dbf9d20): "[lit] Echo full RUN lines in case of external shells".**

Before <https://reviews.llvm.org/D154984> and <https://reviews.llvm.org/D156954>, lit reported full RUN lines in a `Script:` section.  Now, in the case of lit's internal shell, it's the execution trace that includes them.  However, if lit is configured to use an external shell (e.g., bash, windows `cmd`), they aren't reported at all.

A fix was requested at the following:

* <https://reviews.llvm.org/D154984#4627605>
* <https://discourse.llvm.org/t/rfc-improving-lits-debug-output/72839/35?u=jdenny-ornl>

This patch does not address the case when the external shell is windows `cmd`.  As discussed at <https://github.com/llvm/llvm-project/pull/65242>, it's not clear whether that's a use case that people still care about, and it seems to be generally broken anyway.

jdenny-ornl commented:
The current PR replaces [PR #65267](https://github.com/llvm/llvm-project/pull/65267) but has some notable differences:

- I have rebased it onto the most recent versions of [D154984](https://reviews.llvm.org/D154984) and [D156954](https://reviews.llvm.org/D156954), which are based on recent main.
- I have folded in the [fix for the bot failures](https://github.com/llvm/llvm-project/pull/65267#issuecomment-1709171870) that required me to revert the original PR.  I also added a lit test for that case.
- I added a solution for suppressing duplicate printing of RUN lines (once for tracing the associated echo command, and once for the output of that echo command).  It's [proposal 3 in the RFC comment that addresses this issue](https://discourse.llvm.org/t/rfc-improving-lits-debug-output/72839/44?u=jdenny-ornl).  I decided against proposal 1 because later comments expressed concern about how much info would be lost under it.
- I applied python formatting suggestions from darker.

RoboTux reviewed (approved):
LGTM. Note: I've only looked at third commit as instructed, that is: https://github.com/llvm/llvm-project/pull/66408/commits/a398f3e1dfbfc83489e4d85fd74e8b3f12d469e9

jdenny-ornl commented:
> LGTM.

Thanks for the quick review!

> Note: I've only looked at third commit as instructed, that is: [a398f3e](https://github.com/llvm/llvm-project/commit/a398f3e1dfbfc83489e4d85fd74e8b3f12d469e9)

You mean d71c72c9f20ed34dc2fe2d0ec849a6e44dbf9d20, right?  a398f3e1dfbfc83489e4d85fd74e8b3f12d469e9 is the first commit.

RoboTux commented:
> > LGTM.
> 
> Thanks for the quick review!
> 
> > Note: I've only looked at third commit as instructed, that is: [a398f3e](https://github.com/llvm/llvm-project/commit/a398f3e1dfbfc83489e4d85fd74e8b3f12d469e9)
> 
> You mean [d71c72c](https://github.com/llvm/llvm-project/commit/d71c72c9f20ed34dc2fe2d0ec849a6e44dbf9d20), right? [a398f3e](https://github.com/llvm/llvm-project/commit/a398f3e1dfbfc83489e4d85fd74e8b3f12d469e9) is the first commit.

Sadly not. I never remember if top is older or newer. Sorry, I'll review it now then. Please ignore my approval for now.

RoboTux reviewed (approved):
LGTM

jdenny-ornl commented:
> > > LGTM.
> > 
> > 
> > Thanks for the quick review!
> > > Note: I've only looked at third commit as instructed, that is: [a398f3e](https://github.com/llvm/llvm-project/commit/a398f3e1dfbfc83489e4d85fd74e8b3f12d469e9)
> > 
> > 
> > You mean [d71c72c](https://github.com/llvm/llvm-project/commit/d71c72c9f20ed34dc2fe2d0ec849a6e44dbf9d20), right? [a398f3e](https://github.com/llvm/llvm-project/commit/a398f3e1dfbfc83489e4d85fd74e8b3f12d469e9) is the first commit.
> 
> Sadly not. I never remember if top is older or newer. Sorry, I'll review it now then. Please ignore my approval for now.

Sorry for the confusion, and thanks again for your help!  My takeaway is that, for situations like this, I should link to the actual commits to avoid the ambiguity.

RoboTux commented:
> > > > LGTM.
> > > 
> > > 
> > > Thanks for the quick review!
> > > > Note: I've only looked at third commit as instructed, that is: [a398f3e](https://github.com/llvm/llvm-project/commit/a398f3e1dfbfc83489e4d85fd74e8b3f12d469e9)
> > > 
> > > 
> > > You mean [d71c72c](https://github.com/llvm/llvm-project/commit/d71c72c9f20ed34dc2fe2d0ec849a6e44dbf9d20), right? [a398f3e](https://github.com/llvm/llvm-project/commit/a398f3e1dfbfc83489e4d85fd74e8b3f12d469e9) is the first commit.
> > 
> > 
> > Sadly not. I never remember if top is older or newer. Sorry, I'll review it now then. Please ignore my approval for now.
> 
> Sorry for the confusion, and thanks again for your help! My takeaway is that, for situations like this, I should link to the actual commits to avoid the ambiguity.

Or just give the commit title. And only temporarily, I'm sure we'll all be well acquainted with github interface eventually.

jdenny-ornl commented:
> Or just give the commit title. And only temporarily, I'm sure we'll all be well acquainted with github interface eventually.

OK, did that too, in case any one else wants to take a look.  I'll pull the note out when landing as it won't mean much in the final commit log.

Instead of landing on a Friday, I think it will be best to wait until at least early next week.  That will give more time for others to comment, and it will make it easier to respond to any immediate issues.  That seems important given the trouble this series had when I tried to land it last time.

Thanks again for your help in moving this along.

jdenny-ornl committed:
[lit] Echo full RUN lines in case of external shells

Before <https://reviews.llvm.org/D154984> and
<https://reviews.llvm.org/D156954>, lit reported full RUN lines in a
`Script:` section.  Now, in the case of lit's internal shell, it's the
execution trace that includes them.  However, if lit is configured to
use an external shell (e.g., bash, windows `cmd`), they aren't
reported at all.

A fix was requested at the following:

* <https://reviews.llvm.org/D154984#4627605>
* <https://discourse.llvm.org/t/rfc-improving-lits-debug-output/72839/35?u=jdenny-ornl>

This patch does not address the case when the external shell is
windows `cmd`.  As discussed at
<https://github.com/llvm/llvm-project/pull/65242>, it's not clear
whether that's a use case that people still care about, and it seems
to be generally broken anyway.

dyung commented:
There seems to be a race condition of sorts which has caused the test shtest-run-at-line.py to be a bit unreliable on a Windows bot:
https://lab.llvm.org/buildbot/#/builders/216

It seems like there is some extra output that is not expected that is causing the CHECK-NEXT to fail on several lines. For example:
```
             6: Command Output (stderr): 
 check:17       ^~~~~~~~~~~~~~~~~~~~~~~
             7: -- 
 next:18        ^~
             8: bash.exe: warning: could not find /tmp, please create! 
             9: RUN: at line 4: true 
 next:19        !~~~~~~~~~~~~~~~~~~~  error: match on wrong line
```
The extra line `bash.exe: warning: could not find /tmp, please create!` seems to sometimes cause the test to fail. Can the test be restructured such that it can handle this?

jdenny-ornl commented:
> There seems to be a race condition of sorts which has caused the test shtest-run-at-line.py to be a bit unreliable on a Windows bot: https://lab.llvm.org/buildbot/#/builders/216

Thanks for pointing that out.  (And thanks for enabling FileCheck verbosity on this bot!)

>              8: bash.exe: warning: could not find /tmp, please create! 

Do you know why this is being printed at all?  Is it reasonable to just create that directory on that bot?

> Can the test be restructured such that it can handle this?

I'm concerned about weakening the test and missing other problems.  I might use `grep -v` to throw away that line.  Is that portable to windows?

dyung commented:
> > There seems to be a race condition of sorts which has caused the test shtest-run-at-line.py to be a bit unreliable on a Windows bot: https://lab.llvm.org/buildbot/#/builders/216
> 
> Thanks for pointing that out. (And thanks for enabling FileCheck verbosity on this bot!)

Thanks for telling me how to, it is much more useful now!

> > ```
> >          8: bash.exe: warning: could not find /tmp, please create! 
> > ```
> 
> Do you know why this is being printed at all? Is it reasonable to just create that directory on that bot?

I could, although I vaguely recall having similar problems in the past trying to get rid of this error, but I can try. I would prefer that the test can handle it just in case though.

> > Can the test be restructured such that it can handle this?
> 
> I'm concerned about weakening the test and missing other problems. I might use `grep -v` to throw away that line. Is that portable to windows?

I'm not sure. I think there are some tests that still do use grep, but I think the usage of it is highly discouraged these days.

jdenny-ornl commented:
> Thanks for telling me how to, it is much more useful now!

Great!

> > > ```
> > >          8: bash.exe: warning: could not find /tmp, please create! 
> > > ```
> > 
> > 
> > Do you know why this is being printed at all? Is it reasonable to just create that directory on that bot?
> 
> I could, although I vaguely recall having similar problems in the past trying to get rid of this error, but I can try. I would prefer that the test can handle it just in case though.

Understood.

> > > Can the test be restructured such that it can handle this?
> > 
> > 
> > I'm concerned about weakening the test and missing other problems. I might use `grep -v` to throw away that line. Is that portable to windows?
> 
> I'm not sure. I think there are some tests that still do use grep, but I think the usage of it is highly discouraged these days.

I grepped for grep and found a number of tests still using it.  I'll push a commit with this solution for now so that (hopefully) the bot is green again.  We can try to find a better solution afterward.  Does that seem reasonable?

dyung commented:
> > > > Can the test be restructured such that it can handle this?
> > > 
> > > 
> > > I'm concerned about weakening the test and missing other problems. I might use `grep -v` to throw away that line. Is that portable to windows?
> > 
> > 
> > I'm not sure. I think there are some tests that still do use grep, but I think the usage of it is highly discouraged these days.
> 
> I grepped for grep and found a number of tests still using it. I'll push a commit with this solution for now so that (hopefully) the bot is green again. We can try to find a better solution afterward. Does that seem reasonable?

Sure, that sounds reasonable to me.

mstorsjo commented:
> > I'm concerned about weakening the test and missing other problems. I might use `grep -v` to throw away that line. Is that portable to windows?
> 
> I'm not sure. I think there are some tests that still do use grep, but I think the usage of it is highly discouraged these days.

I guess it depends on which testsuite it is. The main llvm testsuite (and the other ones that use the same base setup) rely on a handful of unix-like tools, and will implicitly try to find them from an installation of Git for Windows, unless otherwise available: https://github.com/llvm/llvm-project/blob/llvmorg-17.0.1/llvm/utils/lit/lit/llvm/config.py#L29-L34

jdenny-ornl commented:
This is lit's own test suite, which invokes that same `LLVMConfig` constructor.

jdenny-ornl commented:
> > > > > Can the test be restructured such that it can handle this?
> > > > 
> > > > 
> > > > I'm concerned about weakening the test and missing other problems. I might use `grep -v` to throw away that line. Is that portable to windows?
> > > 
> > > 
> > > I'm not sure. I think there are some tests that still do use grep, but I think the usage of it is highly discouraged these days.
> > 
> > 
> > I grepped for grep and found a number of tests still using it. I'll push a commit with this solution for now so that (hopefully) the bot is green again. We can try to find a better solution afterward. Does that seem reasonable?
> 
> Sure, that sounds reasonable to me.

Pushed that: aa71680f2948ac177144e1089c58d55b9ac0cef2.  I'll keep an eye on the bots for any additional issues.

jdenny-ornl commented:
Fixed another instance of that: 30d77fb80857e645b300c8f59cad9414d090e083

dyung commented:
> Fixed another instance of that: [30d77fb](https://github.com/llvm/llvm-project/commit/30d77fb80857e645b300c8f59cad9414d090e083)

Thanks for the quick fix, was just about to comment on that! Hopefully that's the last of them!

jdenny-ornl commented:
I think we're past that problem now.  The remaining failure appears to be a `/` vs. `\` issue in a new test introduced by [D150987](https://reviews.llvm.org/D150987).  (It also happens to be related to `/tmp`, but I think that's a coincidence.)

dyung commented:
> I think we're past that problem now. The remaining failure appears to be a `/` vs. `\` issue in a new test introduced by [D150987](https://reviews.llvm.org/D150987). (It also happens to be related to `/tmp`, but I think that's a coincidence.)

Yeah, I think so as well. I'm testing a fix for that issue.
